### ISSUE TITLE
[Discuss] port k8s client to kubernetes/client-go

### ISSUE THREAD

**Issue by walktall** [elastic/beats#6068]
Just like what is described in https://github.com/ericchiang/k8s#should-i-use-this-or-client-go.

I think that using kubernetes/client-go is a better choice since caching in libbeat can be handled by client-go Informer directly with less code and better abstract.

@exekias Any idea? I am currently writing code for https://github.com/elastic/beats/issues/5932, and I am thinking about using client-go instead.

**Comment by exekias** [comment]
Hi @walktall, thank you for your input, it's very valid.

While we would like to have all the benefits of `client-go`, it would increase our resulting binary size by more than 30MB. While one could argue it's not that much, Beats support lots of different use cases and hardware, and in some of them, size matters.

We have been trying to avoid this change, at least for now. And from our experience so far, current client is working fine. Probably we can reevaluate this in some months, and check if `client-go` got slimmer or kubernetes use case of beats grew significantly.

**Comment by walktall** [comment]
@exekias Agree with you that size could be matter, I think it is worth to try though. I will try to replace it to client-go first since code change not that much and client-go may not take that much space now than 7 month ago when k8s client vendored.

**Comment by ruflin** [comment]
@walktall I'm quite interested in the outcome of this experiment. I remember doing something similar about a year ago with docker and the outcome on our side was a fork of the docker repo with 1-2 methods removed. The reasons for this was that removing these 1-2 methods (which we didn't need) removed 95% of the dependencies from our side. As far as I remember the dependencies were related to command line options for windows but as we don't access it through the command line, we didn't need all of this. I'm writing this because I could see similar things happening on the k8s-client side.

**Comment by walktall** [comment]
@exekias You are right, I have made a client-go version of beats using latest client-go^6.0.0, for metricbeat, it grows 32MB.

@ruflin I am going to find out if there is something similar to remove in client-go.
